### ISSUE TITLE
✔️ 🚀 Feature Request: Linkable column

### ISSUE THREAD

**Issue by tijmenbrouwers** [elementor/elementor#351]
I am creating a project in Elementor with a lot of "cards". I would like to be able to make the whole "card" clickable. The cards are just columns with some content. Just adding a link to the column element would fix my problem. Now I'm only able to add links to elements within the columns like images and titles.  

![screen shot 2016-08-01 at 13 52 18](https://cloud.githubusercontent.com/assets/3199301/17293200/34c566d4-57ef-11e6-8685-f67e933d3037.png)

For a live example check:
http://winterscout.nl/

**Comment by yehudah** [comment]
Hi,

We thinking about a thing like this because there is a few issues can happen like: semantic, validation, etc...

I have no time frame if it will happen soon.

**Comment by arielk** [comment]
Sorry about the delay. We checked it recently again and unfortunately this option is not feasible using Elementor, so I close this issue for now.

Thanks for the feedback!

**Comment by ThrudUK** [comment]
This is possible using jquery,

**Comment by Pasdoe** [comment]
Any idea on how we can do this using jQuery? A help article for this would be really helpful.
Also, could you go into detail as to why this is not "feasible" within Elementor? Maybe someone else has an idea on how to get around your issue. Just my input.

**Comment by Pasdoe** [comment]
So as I am having a bit of spare time, I came up with a quick and easy temporary solution for creating clickable columns. It is dependant on you using a link within the column, but otherwise it's fairly straightforward. I created a video for everyone wanting to use this method.
Sorry for the crappy microfone, but I had no other one around just now.

Here you can find a quick tutorial:

URL: https://www.useloom.com/share/33ca29be881f4298b4c752c17526e3d9
PW: clickable

If you just want to download a layout and use the instructions there without watching the video, just go ahead and download the layout and import it:

URL: http://share.raidpixels.com/2F2Y3U473J3o
PW: clickable

Hope this helps.

**Comment by ideasandpixels** [comment]
Here's an easy jQuery fix for anyone having trouble. Just give your column a class of "click-child", then add this code: 

`jQuery(document).ready(function() {

	jQuery("body").on("click", ".click-child", function() {

		var href = jQuery(this).find("a").attr('href');

		window.location = href;

	});

});`

**Comment by torndownunit** [comment]
@Pasdoe thanks for your tip.  Out of all the solutions in the various threads, this one seems to look the best due to the extra code for styling the cursor.

It's ridiculous the developers won't address this issue.  Every other builder has this obvious feature.  They say it's not possible, yet there are 3 different solutions posted throughout these threads.  This was my primary reason for not buying the Pro version.  If the developer doesn't care enough to address this, then I don't trust giving them money.

**Comment by alipiogabriel** [comment]
I think its better if you put :not in the body object to prevent redirect when editing the block.

```
jQuery(document).ready(function() {
 	jQuery("body**:not(.elementor-editor-active)**").on("click", ".click-child", function() { 
 		var href = jQuery(this).find("a").attr('href'); window.location = href; 
 	}); 
});
```

**Comment by black-eye** [comment]
+1

**Comment by Rayken** [comment]
Why is this closed? Just add anchors to the HTML-tag option and if anchors are selected, let us provide a custom destination or pick one and a checkbox to open in new tab or not. I don't see how this isn't possible at all. Make the wrapper an anchor, add href and that's that?

**Comment by philmatt** [comment]
+1

**Comment by dmrowell** [comment]
How is it possible this has been described as impossible to do, when there's an excellent plug-in already available on Wordpress.

https://wordpress.org/plugins/make-column-clickable-elementor/ 

Come on, Elementor.  If you don't want to be swamped by Gutenberg, you can do better than to describe something which has already been solved as being impossible.  Can you please simply add this to your core code.

**Comment by mredodos** [comment]
+1

**Comment by klainser** [comment]
@arielk how this feature is not possible since there's a Elementor module for it ? 

https://wordpress.org/plugins/make-column-clickable-elementor

**Comment by chloeford0** [comment]
+1 

This would be so helpful and allow much more flexibility.

**Comment by JulimanS** [comment]
Hi guys, thanks a lot for your pro-activity and patch/solutions to that problem.

Your solutions works very well in case you have set manually X columns in Elementor but I think is not solving the problem @tijmenbrouwers is describing, due that he is using the Archive Post Widget and Cards.

When you include the Archive Post in a widget, you can not select and point the different columns in order to include the .class. Everything is an entire section from the Elementor Edition Screen.

I have tested your codes and solutions (@pascalinger , @alipiogabriel @ideasandpixels ) and in the case of using the Archive Post Widget, all cards points to the same link (in this case, the link of the first article / card you have listed).

Do anyone be so kind to invest some time how can we apply this solution to the archive post widget?

Thanks a lot!

**Comment by torndownunit** [comment]
> Hi guys, thanks a lot for your pro-activity and patch/solutions to that problem.
> 
> Your solutions works very well in case you have set manually X columns in Elementor but I think is not solving the problem @tijmenbrouwers is describing, due that he is using the Archive Post Widget and Cards.
> 
> When you include the Archive Post Widget in a section, you can not select and point the different columns in order to include the .class. Everything is an entire section from the Elementor Edition Screen.
> 
> I have tested your codes and solutions (@pascalinger , @alipiogabriel @ideasandpixels ) and in the case of using the Archive Post Widget, all cards/articles/post point to the same link (in this case, the link of the first article / card you have in the archive list).
> 
> Do anyone be so kind to invest some time how can we apply this solution to the archive post widget?
> 
> Thanks a lot!

This is a completely different issue and should be filed as such. Making columns clickable, and changing how a whole widget works are separate issues.  I agree with you, the functionality would be fantastic.  But if it's filed as a separate issue, it might actually get some notice.  Though Elementor clearly don't care about this issue since this has been active since 2016 with no resolution besides fixes users have posted.  So likely not.

**Comment by torndownunit** [comment]
> Why is this closed? Just add anchors to the HTML-tag option and if anchors are selected, let us provide a custom destination or pick one and a checkbox to open in new tab or not. I don't see how this isn't possible at all. Make the wrapper an anchor, add href and that's that?

Not only do they keep closing this issue even though it's very active, there are frequently other requests for this posted, and they close those and merge them with this, another post they have closed.  It's ridiculous.

**Comment by JulimanS** [comment]
Hi @torndownunit , thanks for your response.

In fact, the issue I describe in my message is the exact same issue the autor's thread is looking for. The solutions provided, even they are super useful to make columns clickable, they are not addressing the problem the author is describing.

So, now that is thread is well positioned, the correct approach from my point of view should solving the problem in this thread and not duplicate it.

**Comment by torndownunit** [comment]
> Hi @torndownunit , thanks for your response.
> 
> In fact, the issue I describe in my message is the exact same issue the autor's thread is looking for. The solutions provided, even they are super useful to make columns clickable, they are not addressing the problem the author is describing.
> 
> So, now that is thread is well positioned, the correct approach from my point of view should be solving the problem in this thread and not duplicate it.

The developers absolutely do not care about the users of the free version of the plugin.  This thread proves that.  Look how many times they have merged similar issues and then closed this.   So welcome to the thread, but don't expect any activity on it.  This has been going on since 2016.

**Comment by JulimanS** [comment]
I have not mentioned that the solution has to come from the Elementor developers, but from anyone who has the time and the desire to offer a direct solution to what the thread author is really asking, I think it is quite clear.

Either way.

A pretty good coder friend has sent me a solution for the problem of using the entire card as link to the article. The workaround is to use the read more link so that the link now takes up all the space on the card via CSS.

I have tested the code and it works perfectly.

```
.post-type-archive-blog .elementor-post .elementor-post__read-more {
position: absolute;
top: 0;
bottom: 0;
left: 0;
right: 0;
opacity: 0;
    margin-bottom: 0
}
```

I hope it helps!

**Comment by MichaelDarko** [comment]
Being able to apply a link to whole section/column is absolutely essential in website building process.

Elementor markets itself as the best page builder and then they tell us that adding a link to a section (first thing you learn in HTML) will not be possible.

Whatever the reason is, they should solve it. Other page builders have it, Webflow has it. It's not impossible.

But I'm pretty sure something so essential like this will maybe come after 50 other "fancy" features that nobody was asking for.
_____________________________________________

**I'm currently using this annoying workaround:**

**1.** Place a HTML widget in the top of a section/column

**2.** Add a link into the HTML widget

`<a href="https://www.domain.com"></a>`

**3.** Stretch the HTML widget with the link over the whole section/column with custom CSS

(Add this to the HTML widget custom CSS in the advanced tab.)
```
selector a {
    
position: absolute;
height: 100%;
width: 100%;
z-index: 9;    
    
}

selector {
    
position: absolute;
height: 100%;
width: 100%;  
    
}
```
It does the job, but you will have to access everything in the section through the navigator, because the whole section will be covered by the HTML widget in the editor.

It's emabrrassing that we have to use silly workarounds for something like this.

**Comment by ChristopherKrohn** [comment]
The fact that this is not possible is just insane..
I just don't understand how this very simple and obvious feature is not a part of the builder ut of the box. 

I guess I will just have to make silly workarounds like the one above, but this is seriously breaking the "ease of use" that I am selling Elementor on, when setting up websites for them to later make changes to.

Just explaining this workaround, ruins my credibility... The standard response from a client is "Of course it's possible."

Please, please please add this feature to the builder..

**Comment by bouveng** [comment]
It's perfectly valid in HTML5 to wrap the column div with an a-tag but the problem is that if you add other linked content inside the column it will be invalid and won't work.

That might be the major cause of why Elementor not implementing this.

However, the cleanest way this can be done is by using the section custom `attribute` with a `data-href|http://example.com` and a little js somewhere, preferably in a child theme.

`( function( $ ){

    $( '[data-href]' ).click( function(){
        window.location.href = $( this ).attr( 'data-href' );
    });

})( jQuery );`

**Comment by viveknath322** [comment]
> It's perfectly valid in HTML5 to wrap the column div with an a-tag but the problem is that if you add other linked content inside the column it will be invalid and won't work.
> 
> That might be the major cause of why Elementor not implementing this.
> 
> However, the cleanest way this can be done is by using the section custom `attribute` adding a `data-href|http://example.com` and a little js somewhere, preferably in a child theme.
> 
> ```
> ( function( $ ){
> 
>     $( '[data-href]' ).click(function(){
>         window.location.href = $( this ).attr( 'data-href' );
>     });
> 
> })( jQuery );
> ```

This worked perfectly, but how can I change the mouse icon to pinter while hovering?

**Comment by JamieWalkerSTF** [comment]
+1 - It's possible in Divi, but not Elementor?

**Comment by Rayken** [comment]
I've been asking this question as well. I understand the reason they might be hesitant to add this functionality though. Nested anchors are not valid HTML. And it breaks layout. That being said, if you know what you're doing (as in, won't have any anchors within anchors) then this option should absolutely be a thing!

Use the code below if you know what you're doing, is basically what I'm trying to say. Either copy it and put in your functions as is or make a plugin out of it.

What it does:
- adds the option to set the HTML-tag of a column to <a> (in the layout tab)
- if you select <a> as the HTML-tag, a link field appears below
- if you set a url, the entire column will be clickable
- the attributes such as open in new window, nofollow and "custom attributes" should work

Again: make sure you don't have any other linked elements within the column if you plan on using this. 

```php
// add a URL control to the column layout tab, if <a> is selected as html_tag
add_action('elementor/element/column/layout/before_section_end', function( $section, $args ) {
    $section->add_control(
        'column_link',
        [
            'label'         => __( 'Link', 'elementor' ) . ' (' . strtolower( __( 'Column', 'elementor' ) ) . ')',
            'description'   => 'NOTE: Using anchors within this column will break it! Anchors within anchors is not valid HTML.',
            'type'          => \Elementor\Controls_Manager::URL,
            'placeholder'   => __( 'https://your-link.com', 'elementor' ),
            'show_external' => true,
            'label_block'   => true,
            'default'       => [
                'url'           => '',
                'is_external'   => false,
                'nofollow'      => false
            ],
            'condition'     => [
                'html_tag'      => [ 'a' ]
            ]
        ]
    );
}, 11, 2 );

// add <a> to the html_tag control
add_action( 'elementor/element/column/layout/before_section_end', function( $control_stack, $args ) {
    $control = \Elementor\Plugin::instance()->controls_manager->get_control_from_stack( $control_stack->get_unique_name(), 'html_tag' );
    $control['options']['a'] = __( 'a' );
    $control_stack->update_control( 'html_tag', $control );
}, 10, 2 );

// modify output of column to handle <a>
add_action( 'elementor/frontend/column/before_render', function( $widget ) {
    $settings = $widget->get_settings_for_display();
    if( $settings['html_tag'] !== 'a' )
        return;

    if( empty( $settings['column_link']['url'] ) )
        return;

    $widget->add_link_attributes( '_wrapper', $settings['column_link'] );
}, 10 );
```

If you want to make a plugin out of it, essentially paste the above code within:
```php
/**
 * Plugin Name: Elementor Linkable Columns
 * Description: Elementor extension that enables linkable columns.
 * Version:     1.0.0
 * Author:      Rayken
 * Text Domain: elementor-linkable-columns
 */

if( !defined( 'ABSPATH' ) )
    exit;

add_action( 'plugins_loaded', function() {
    if( !did_action( 'elementor/loaded' ) )
        return;
    
    ... CODE HERE ...
});
```

Well that's it. I guess this response is valid for https://github.com/elementor/elementor/issues/1956 as well. Probably a couple of more related "issues".

**Comment by torndownunit** [comment]
@Rayken thanks, I will be trying this out on a dev site soon.

**Comment by Rayken** [comment]
The above "fix" was just patched in 3.1.2 vs. 3.1.1. Specifically it's [this change](https://github.com/elementor/elementor/commit/5b414fdb8f8c1d344c1f45b2afefe446a6be4d13#diff-54dd09806a56523ef550e195aff7fd24e0fdc7e09e0cfcae6a0795444d69ee40).

The change made it so there's a list of allowed "wrapper tags", and it's not filterable for some reason. So, we're back to square one with CSS and/or javascript trickery. Unless they allow us to filter the `ALLOWED_HTML_WRAPPER_TAGS` constant of `utils.php` to allow `<a>` again.

Changing it to something like:
```php
public static function validate_html_tag( $tag ) {
    $allowed_tags = apply_filters( 'elementor/utils/allowed_tags', self::ALLOWED_HTML_WRAPPER_TAGS );
    return in_array( strtolower( $tag ), $allowed_tags ) ? $tag : 'div';
}
```
should do the trick.

I do have an idea for how we could make sure to filter out/replace any "residual" `<a>`-tags within a column if you've made the column itself clickable (to prevent layout breaking). 

Basically either by using the `elementor/widget/render_content` hook with a replace such as `$content = preg_replace( ['"<a (.*?)>"', '"</a>"'], ['',''], $content );` if the widget is a column and has column_link set, or by using DOM document on the `elementor/frontend/the_content` hook instead. That way we'd make sure "no anchors within an anchored column can occur".

@matipojo or @bainternet would you like to comment? Could we get the allowed tags filterable? Or at least consider the above code (https://github.com/elementor/elementor/issues/351#issuecomment-789732511) and this comment? The CSS hacks and javascript tricks aren't really good methods of linking columns, and aren't that great for SEO either. The above approach at least tries to use Elementor "natively" so achieve what so many people are requesting. Just Google 'elementor link column' and you'll see an ocean of requests, hacks and workarounds.

**Comment by JamieWalkerSTF** [comment]
I really appreciate your effort with this issue Rayken!

**Comment by ajoah** [comment]
I made a PR with your code @Rayken to try to speed things up.

**Comment by Rayken** [comment]
The [code above](https://github.com/elementor/elementor/issues/351#issuecomment-789732511) works again. [Here's a gist](https://gist.github.com/Rayken/468b9534c9d84a08efcbda5ac5e0e618) if you prefer. Please note that nested anchor tags within anchor tags will break layout.

**Comment by hipnie** [comment]
Unfortunately it doesn't work on my page.
Do I paste this code into the custom page-css or site-css?

**Comment by Rayken** [comment]
Neither. As stated in the comment, the code is for users who know what they are doing. I wouldn't recommend you using the snippet/gist otherwise. It's more of a "Hey Elementor, it is absolutely possible and here's a way to do it", just as FusionBuilder, Divi and so many other page builders.

If you don't care about my recommendation, go ahead and read the comments more thoroughly.

**Comment by shilo-ey** [comment]
Reopened this issue.

**Comment by Rayken** [comment]
@shilo-ey nice! What's the PR?

**Comment by pascalinger** [comment]
I just saw that a bunch of people mentioned me here. Not an Elementor user anymore but shocked that this simple request has still not been implemented yet.

@Rayken clever work around the limitation. I like it! 👍

**Comment by JacobBannier** [comment]
> Sorry about the delay. We checked it recently again and unfortunately this option is not feasible using Elementor, so I close this issue for now.
> 
> Thanks for the feedback!

Incredible response from a company with over $15m in funding. Simple plugins from guys sitting in their mum's basement have solved this issue for years, why can't this be added?

**Comment by Gabugbuster** [comment]
It is possible to add a link to the upcoming containers (that replace the section/column structure)
<img width="295" alt="Screen Shot 2022-03-02 at 14 41 01" src="https://user-images.githubusercontent.com/95359071/156363420-f828dedc-fac1-4fc0-b4f2-bccd9f5399e3.png">

**Comment by nicholaszein** [comment]
Hello everybody!

We have great news! 🙌 

### 📢 We're happy to announce that the issue you raised was resolved in Elementor **v3.8.0**! 🥳 

You can now use the (now stable) **Flexbox Container** experiment, and change the HTML tag to an anchor tag `<a>`!

✔️ Feel free to check it out and update your plugin to the new version!

Cheers 🥂
